src/Parallel/Algorithm.hpp

```diff
@@ -331,31 +401,92 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
   // @}
 
   void start_phase(const PhaseType next_phase) noexcept {
-    // terminate should be true since we exited a phase previously.
-    if (not get_terminate()) {
-      ERROR(
-          "An algorithm must always be set to terminate at the beginning of a "
-          "phase. Since this is not the case the previous phase did not end "
-          "correctly. The integer corresponding to the previous phase is: "
-          << static_cast<int>(phase_)
-          << " and the next phase is: " << static_cast<int>(next_phase));
-    }
-    // set terminate to true if there are no actions in this PDAL
-    set_terminate(number_of_actions_in_phase(next_phase) == 0);
+    const PhaseType previous_phase = phase_;
     phase_ = next_phase;
-    algorithm_step_ = 0;
-    perform_algorithm();
+    bool should_start_phase = false;
+    if constexpr(detail::has_required_phase_sync_function<metavariables>) {
+      should_start_phase =
+          (not static_cast<bool>(return_phase_) or phase_ == *return_phase_ or
```

Comment:
So, the `return_phase_` only gets set when the algorithm is told to stop for the sync; then it represents the phase that will be returned to after the collection of 'sync phases'. So, if the return phase is unset, we are not currently going through the set of sync phases, so we should proceed through the phase list normally (so it accepts the next phase dictated by the main chare).
I agree the logic here can be pretty confusing because of all of the state possibilities when it starts a new phase, so I'll try to clarify in the comments, but perhaps you have some ideas about how this can be clearer with better variable names and documentation elsewhere?